Pagination branch

Divo123-tech opened the pull request:
## Things I did
**Page and Limit Validation**
I put the pagination function in the utils/utils.ts file with input validation that defaults a non-number value for skip to be 0, and limit to be 10

**Sorting**
Added sorting in the items.service.ts file and sorted by the name's alphabetical order

## Things to Review
1.the getPagination function in the utils/utils.ts file
2.the sorting in the items.service.ts file

yharsono commented on src/controller/items.controller.ts:
```diff
@@ -1,5 +1,41 @@
+  static async getAll(req: Request, res: Response): Promise<void> {
+    try {
+        const {page, limit} = req.query
+        const skip: number = (Number(page) - 1) * Number(limit) 
```
what if page and limit is not part of the query?
or what if there is only limit, but no page? or vice versa?

yharsono commented on src/service/item.service.ts:
```diff
@@ -0,0 +1,60 @@
+const getAll = async(skp: number,lmt : number) : Promise<Array<ItemInterface>> => {
+    
+    try {
+        return await itemModel.find({isActive: true}).skip(skp || 0).limit(lmt || 10); 
```
you should sort this by something to make sure you always get consistent result. 
Its better to also tell the FE the total counts of the actual list, so they know how many pages they can show

yharsono commented on src/service/item.service.ts:
```diff
@@ -1,65 +1,67 @@
+      { $skip: skip },
+      { $limit: limit },
+    ]);
+    return items;
```
you are still not returning the count of all the items

Divo123-tech replied:
should I make a new interface to return this kind of object?
{total:  number, data: items[]}

yharsono commented on src/service/item.service.ts:
```diff
@@ -1,65 +1,67 @@
-}
+const add = async (newItem: ItemInterface): Promise<ItemInterface> => {
+  try {
+    // return newItem;
```
nit: remove unused code

yharsono commented on src/utils/utils.ts:
```diff
@@ -1,21 +1,16 @@
+  //if limit is not given as a specific number use 10 as the default value
+  const limit: number = query.limit || 10;
+  const page: number = query.page || 1;
+  //query validation implemented in this function
```
what query validation?

yharsono commented on src/utils/utils.ts:
```diff
@@ -0,0 +1,16 @@
+export interface QueryInterface {
```
this is PaginationInterface, not QueryInterface

yharsono commented on src/utils/utils.ts:
```diff
@@ -0,0 +1,16 @@
+  const limit: number = query.limit || 10;
+  const page: number = query.page || 1;
+  //query validation implemented in this function
+  let skip: number = (page - 1) * limit; // Skip can be a number
```
const skip.

yharsono commented on src/service/item.service.ts:
```diff
@@ -1,59 +1,66 @@
+      { $skip: skip },
+      { $limit: limit },
+    ]);
+    return { totalCount: items.length, data: items };
```
you cant just do items.lenght, cause its already been "limited" jdi this count will always return what the limit is, unless the amounts of items are smaller than the limit.

Check my previous comment on using $facet

Divo123-tech commented:
<h3>Added Multiple Aggregation to getAll function</h3>
used .facet() to avoid running multiple queries and return an object with the total number of items in the list that are active, and all the items according to the skip, limit and sort filters.

<h3>Accidentally Merged services-branch to main </h3>
Merged the changes of services-branch to this branch so the changes can be reviewed
<h4>Adding new item interface for the front end.</h4>
 Added this interface for the add and edit functions to use as inputs to avoid front end editing or adding certain properties.
<h4>Used .save() to activate validation</h4>
removed doesKodeExist as the function became unnecesarry after using .save()
<h4>Used Object methods like .assign and .definePropertry() </h4>
Object.assign() was to update the item, and .defineProperty() was to ensure that isActive can only be made false by delete not edit or add.
<h4>Changed delete route </h4>
Changed delete from accepting id in the body to using the request params.

yharsono commented on src/service/item.service.ts:
```diff
@@ -33,42 +49,42 @@ const add = async (newItem: FEItemInterface): Promise<ItemInterface> => {
-        });
+    //adds isActive property to newItem and if FrontEnd makes isActive into false
+    //it overrides it to true
+    Object.defineProperty(item, "isActive", {
```
I dont think we need to worry about this anymore now that we have the FEItemInterface. 
And you could just do: `itemToUpdate.isActive = true` to always set it to true

yharsono commented on src/service/item.service.ts:
```diff
@@ -33,42 +49,42 @@ const add = async (newItem: FEItemInterface): Promise<ItemInterface> => {
+): Promise<FEItemInterface> => {
+  try {
+    const itemToUpdate = await itemModel.findById(id);
+    if (!itemToUpdate) {
```
you also need to check if `!itemToUpdate.isActive`, then you want to throw the item not found error, cause the item has been deleted

yharsono commented on src/service/item.service.ts:
```diff
@@ -33,42 +49,42 @@ const add = async (newItem: FEItemInterface): Promise<ItemInterface> => {
 
-        //assigns the ite passed to the item to be updated
-        Object.assign(itemToUpdate, item)
+    //assigns the ite passed to the item to be updated
```
nit: typo in the comment
